tools/gfx/renderer-shared.h

```diff
@@ -718,20 +716,57 @@ class ShaderObjectBaseImpl : public ShaderObjectBase
                 SLANG_ASSERT(false && "Unexpected specialization argument kind.");
                 return SLANG_FAIL;
             }
-            m_userProvidedSpecializationArgs->add(extendedType);
+            list.add(extendedType);
         }
         return SLANG_OK;
     }
 
-    // Appends all types that are used to specialize the element type of this shader object in
-    // `args` list.
-    virtual Result collectSpecializationArgs(ExtendedShaderObjectTypeList& args) override
+    virtual SLANG_NO_THROW Result SLANG_MCALL setSpecializationArgs(
+        ShaderOffset const& offset,
+        const slang::SpecializationArg* args,
+        uint32_t count) override
     {
-        if (m_userProvidedSpecializationArgs)
+        auto layout = getLayout();
+
+        // If the shader object is a container, delegate the processing to
+        // `setSpecializationArgsForContainerElements`.
+        if (layout->getContainerType() != ShaderObjectContainerType::None)
         {
-            args.addRange(*m_userProvidedSpecializationArgs);
+            ExtendedShaderObjectTypeList argList;
+            SLANG_RETURN_ON_FAIL(
+                getExtendedShaderTypeListFromSpecializationArgs(argList, args, count));
+            setSpecializationArgsForContainerElement(argList);
             return SLANG_OK;
         }
+
+        if (offset.bindingRangeIndex < 0)
+            return SLANG_E_INVALID_ARG;
+        if (offset.bindingRangeIndex >= layout->getBindingRangeCount())
+            return SLANG_E_INVALID_ARG;
+
+        auto bindingRangeIndex = offset.bindingRangeIndex;
+        auto bindingRange = layout->getBindingRange(bindingRangeIndex);
+        auto objectIndex = bindingRange.subObjectIndex + offset.bindingArrayIndex;
+        if (objectIndex >= m_userProvidedSpecializationArgs.getCount())
+            m_userProvidedSpecializationArgs.setCount(objectIndex + 1);
+        if (!m_userProvidedSpecializationArgs[objectIndex])
+        {
+            m_userProvidedSpecializationArgs[objectIndex] =
```

Comment:
I'm not clear here why `m_userProvidedSpecializationArgs` is indexed by the `objectIndex`. It seems that for a sub-object range that uses an array:

```
struct MyParams
{
    ...
    IThing things[10];
}
```

there would only be a single specialization argument needed/wanted for the `things` array.

It almost seems like this is a lurking bug in other parts of the specialization logic: we may be acting as if each *object* in a sub-object range yields a distinct specialization parameter, while the reality is that the entire *range* will have a single specialization parameter (and thus a single argument as well).

It seems that in the long run it will be simplest if each sub-object range pre-computes a "flat" index into a single array of specialization arguments related to that range, much like each binding range stores a flat index like `bindingRange.subObjectIndex` here.

Reply:
You are right, this is likely wrong for the array case. However I am not getting into array support yet, so I guess I can let it slip a little bit right now until I need to make it work for arrays.

Reply:
I should also note that in the case of an array, this code branch will not be executed, and the code where your previous comment is will be used instead, which should be correct.